### PR TITLE
ffmpeg: add additional nvidia options

### PR DESCRIPTION
**Pull request by jopejoe1**
## Description of changes
Add options to uses Nvidia proprietary cuda compiler in ffmpeg, for testing an nvidia GPU is needed, ffmpeg needs to be set to enable unfree and cuda must be enabled.

```nix
ffmpeg-full.override {
  config = {
    cudaSupport = true;
  };
  withUnfree = true;
}
```

The resulting ffmpeg binary can then be tested with the following 2 commands:

```bash
ffmpeg -hwaccel cuda -f lavfi -i testsrc -t 10  -vf "format=nv12,hwupload_cuda,scale_cuda=1280:720" -c:v h264_nvenc /tmp/cuda-nvcc-test.mkv
ffmpeg -hwaccel cuda -f lavfi -i testsrc -t 10  -vf "format=nv12,hwupload_cuda,scale_npp=1280:720" -c:v h264_nvenc /tmp/npp-test.mkv
```

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/libraries/ffmpeg/generic.nix [hunk 1]
```diff
@@ -1,4 +1,4 @@
-{ lib, stdenv, buildPackages, removeReferencesTo, addDriverRunpath, pkg-config, perl, texinfo, texinfo6, yasm
+{ lib, config, stdenv, buildPackages, removeReferencesTo, addDriverRunpath, pkg-config, perl, texinfo, texinfo6, yasm
```
Comment (justinas) [inline pkgs/development/libraries/ffmpeg/generic.nix:1]: Many packages have a `cudaSupport ? config.cudaSupport`, i.e. `cudaSupport` is also a top-level parameter. Not sure which pattern is preferred.

**Review by justinas — APPROVED** [review]
Tested using the given commands, seems to work (NixOS 24.05, Nvidia driver 550.78, RTX 4070).

File: pkgs/development/libraries/ffmpeg/generic.nix (hunk 1 above)
Comment (jopejoe1) [reply pkgs/development/libraries/ffmpeg/generic.nix:1]: I think adding that as an option would be a bit confusing, as we already support some cuda features on our full FFmpeg version.

Comment (justinas) [reply pkgs/development/libraries/ffmpeg/generic.nix:1]: True, usually "enable CUDA support" is a binary switch, but ffmpeg can use CUDA with only free components, or with unfree ones. :confused:

**Review by Atemu — APPROVED** [review]
I can't test this but diff LGTM.

**Pull request merged**